File: src/API/ontologenius/ConversionClient.cpp
#include "ontologenius/API/ontologenius/ConversionClient.h"

#include <cstdint>
#include <iostream>
#include <string>
#include <vector>

#include "ontologenius/compat/ros.h"

#ifndef COLOR_OFF
#define COLOR_OFF "\x1B[0m"
#endif
#ifndef COLOR_RED
#define COLOR_RED "\x1B[0;91m"
#endif
#ifndef COLOR_ORANGE
#define COLOR_ORANGE "\x1B[1;33m"
#endif
#ifndef COLOR_GREEN
#define COLOR_GREEN "\x1B[1;92m"
#endif

namespace onto {

  using namespace ontologenius;
  using OntologeniusConversionRequestAlias = compat::RawRequestType<ontologenius::compat::OntologeniusConversion>;

  ConversionClient::ConversionClient(const std::string& name) : name_((name.empty()) ? "/ontologenius/conversion" : "/ontologenius/conversion/" + name),
                                                                client_verbose_(false),
                                                                client_(name_)
  {}

  std::vector<std::string> ConversionClient::individualsIndex2Id(const std::vector<int64_t>& indexes)
  {
    return index2Id(indexes, OntologeniusConversionRequestAlias::INDIVIDUALS);
  }

  std::vector<std::string> ConversionClient::classesIndex2Id(const std::vector<int64_t>& indexes)
  {
    return index2Id(indexes, OntologeniusConversionRequestAlias::CLASSES);
  }

  std::vector<std::string> ConversionClient::dataPropertiesIndex2Id(const std::vector<int64_t>& indexes)
  {
    return index2Id(indexes, OntologeniusConversionRequestAlias::DATA_PROPERTIES);
  }

  std::vector<std::string> ConversionClient::objectPropertiesIndex2Id(const std::vector<int64_t>& indexes)
  {
    return index2Id(indexes, OntologeniusConversionRequestAlias::OBJECT_PROPERTIES);
  }

  std::vector<std::string> ConversionClient::literalsIndex2Id(const std::vector<int64_t>& indexes)
  {
    return index2Id(indexes, OntologeniusConversionRequestAlias::LITERAL);
  }

  std::string ConversionClient::individualsIndex2Id(int64_t index)
  {
    return index2Id(index, OntologeniusConversionRequestAlias::INDIVIDUALS);
  }

  std::string ConversionClient::classesIndex2Id(int64_t index)
  {
    return index2Id(index, OntologeniusConversionRequestAlias::CLASSES);
  }

  std::string ConversionClient::dataPropertiesIndex2Id(int64_t index)
  {
    return index2Id(index, OntologeniusConversionRequestAlias::DATA_PROPERTIES);
  }

  std::string ConversionClient::objectPropertiesIndex2Id(int64_t index)
  {
    return index2Id(index, OntologeniusConversionRequestAlias::OBJECT_PROPERTIES);
  }

  std::string ConversionClient::literalsIndex2Id(int64_t index)
  {
    return index2Id(index, OntologeniusConversionRequestAlias::LITERAL);
  }

  std::vector<int64_t> ConversionClient::individualsId2Index(const std::vector<std::string>& ids)
  {
    return id2Index(ids, OntologeniusConversionRequestAlias::INDIVIDUALS);
  }

  std::vector<int64_t> ConversionClient::classesId2Index(const std::vector<std::string>& ids)
  {
    return id2Index(ids, OntologeniusConversionRequestAlias::CLASSES);
  }

  std::vector<int64_t> ConversionClient::dataPropertiesId2Index(const std::vector<std::string>& ids)
  {
    return id2Index(ids, OntologeniusConversionRequestAlias::DATA_PROPERTIES);
  }

  std::vector<int64_t> ConversionClient::objectPropertiesId2Index(const std::vector<std::string>& ids)
  {
    return id2Index(ids, OntologeniusConversionRequestAlias::OBJECT_PROPERTIES);
  }

  std::vector<int64_t> ConversionClient::literalsId2Index(const std::vector<std::string>& ids)
  {
    return id2Index(ids, OntologeniusConversionRequestAlias::LITERAL);
  }

  int64_t ConversionClient::individualsId2Index(const std::string& id)
  {
    return id2Index(id, OntologeniusConversionRequestAlias::INDIVIDUALS);
  }

  int64_t ConversionClient::classesId2Index(const std::string& id)
  {
    return id2Index(id, OntologeniusConversionRequestAlias::CLASSES);
  }

  int64_t ConversionClient::dataPropertiesId2Index(const std::string& id)
  {
    return id2Index(id, OntologeniusConversionRequestAlias::DATA_PROPERTIES);
  }

  int64_t ConversionClient::objectPropertiesId2Index(const std::string& id)
  {
    return id2Index(id, OntologeniusConversionRequestAlias::OBJECT_PROPERTIES);
  }

  int64_t ConversionClient::literalsId2Index(const std::string& id)
  {
    return id2Index(id, OntologeniusConversionRequestAlias::LITERAL);
  }

  std::vector<std::string> ConversionClient::index2Id(const std::vector<int64_t>& indexes, int8_t source)
  {
    auto req = ontologenius::compat::makeRequest<ontologenius::compat::OntologeniusConversion>();
    auto res = ontologenius::compat::makeResponse<ontologenius::compat::OntologeniusConversion>();

    [source, indexes](auto&& req) {
      req->source = source;
      req->values_int = indexes;
    }(ontologenius::compat::onto_ros::getServicePointer(req));

    if(call(req, res))
      return [](auto&& res) {
        return res->values_str;
      }(ontologenius::compat::onto_ros::getServicePointer(res));
    else
      return {};
  }

  std::string ConversionClient::index2Id(int64_t index, int8_t source)
  {
    return index2Id(std::vector<int64_t>{index}, source)[0];
  }

  std::vector<int64_t> ConversionClient::id2Index(const std::vector<std::string>& ids, int8_t source)
  {
    auto req = ontologenius::compat::makeRequest<ontologenius::compat::OntologeniusConversion>();
    auto res = ontologenius::compat::makeResponse<ontologenius::compat::OntologeniusConversion>();

    [source, ids](auto&& req) {
      req->source = source;
      req->values_str = ids;
    }(ontologenius::compat::onto_ros::getServicePointer(req));

    if(call(req, res))
      return [](auto&& res) {
        return res->values_int;
      }(ontologenius::compat::onto_ros::getServicePointer(res));
    else
      return {};
  }

  int64_t ConversionClient::id2Index(const std::string& id, int8_t source)
  {
    return id2Index(std::vector<std::string>{id}, source)[0];
  }

  bool ConversionClient::call(compat::RequestType<compat::OntologeniusConversion>& req,
                              compat::ResponseType<compat::OntologeniusConversion>& res)
  {
    using ResultTy = typename decltype(client_)::Status_e;

    switch(client_.call(req, res))
    {
    case ResultTy::ros_status_successful_with_retry:
      [[fallthrough]];
    case ResultTy::ros_status_successful:
    {
      return true;
    }
    case ResultTy::ros_status_failure:
    {
      if(client_verbose_)
        std::cout << COLOR_RED << "Failed to call ontologenius/" << name_ << COLOR_OFF << std::endl;
      [[fallthrough]];
    }
    default:
      return false;
    }
  }

} // namespace onto

File: ontopy/ontologenius/ConversionClient.py
from .compat.ros import Ontoros, OntoService
import os

from ontologenius.srv import OntologeniusConversion
if os.environ["ROS_VERSION"] == "1":
    from ontologenius.srv import OntologeniusConversionRequest
else:
    from ontologenius.srv._ontologenius_conversion import OntologeniusConversion_Request as OntologeniusConversionRequest

class ConversionClient:

    def __init__(self, name):
        """Constructs a ROS client linked to the service name(str)."""
        self._name = 'ontologenius/conversion'
        if name != '':
          self._name = self._name + "/" + name
        self._verbose = False
        self._client = Ontoros.createService(self._name, OntologeniusConversion)

    def setVerbose(self, verbose):
        """If verbose(bool) is set to True, the clients will post messages about
           the failure to call the services and about their restoration.
        """
        self._verbose = verbose

    def individualsIndexes2Ids(self, indexes):
       return self._indexes2Ids(indexes, OntologeniusConversionRequest.INDIVIDUALS)

    def classesIndexes2Ids(self, indexes):
       return self._indexes2Ids(indexes, OntologeniusConversionRequest.CLASSES)

    def dataPropertiesIndexes2Ids(self, indexes):
       return self._indexes2Ids(indexes, OntologeniusConversionRequest.DATA_PROPERTIES)

    def objectPropertiesIndexes2Ids(self, indexes):
       return self._indexes2Ids(indexes, OntologeniusConversionRequest.OBJECT_PROPERTIES)

    def literalsIndexes2Ids(self, indexes):
       return self._indexes2Ids(indexes, OntologeniusConversionRequest.LITERAL)
    

    def individualsIndex2Id(self, index):
       return self._index2Id(index, OntologeniusConversionRequest.INDIVIDUALS)

    def classesIndex2Id(self, index):
       return self._index2Id(index, OntologeniusConversionRequest.CLASSES)

    def dataPropertiesIndex2Id(self, index):
       return self._index2Id(index, OntologeniusConversionRequest.DATA_PROPERTIES)

    def objectPropertiesIndex2Id(self, index):
       return self._index2Id(index, OntologeniusConversionRequest.OBJECT_PROPERTIES)

    def literalsIndex2Id(self, index):
       return self._index2Id(index, OntologeniusConversionRequest.LITERAL)
    

    def individualsIds2Indexes(self, ids):
       return self._ids2Indexes(ids, OntologeniusConversionRequest.INDIVIDUALS)

    def classesIds2Indexes(self, ids):
       return self._ids2Indexes(ids, OntologeniusConversionRequest.CLASSES)

    def dataPropertiesIds2Indexes(self, ids):
       return self._ids2Indexes(ids, OntologeniusConversionRequest.DATA_PROPERTIES)

    def objectPropertiesIds2Indexes(self, ids):
       return self._ids2Indexes(ids, OntologeniusConversionRequest.OBJECT_PROPERTIES)

    def literalsIds2Indexes(self, ids):
       return self._ids2Indexes(ids, OntologeniusConversionRequest.LITERAL)
    

    def individualsId2Index(self, id):
       return self._id2Index(id, OntologeniusConversionRequest.INDIVIDUALS)

    def classesId2Index(self, id):
       return self._id2Index(id, OntologeniusConversionRequest.CLASSES)

    def dataPropertiesId2Index(self, id):
       return self._id2Index(id, OntologeniusConversionRequest.DATA_PROPERTIES)

    def objectPropertiesId2Index(self, id):
       return self._id2Index(id, OntologeniusConversionRequest.OBJECT_PROPERTIES)

    def literalsId2Index(self, id):
       return self._id2Index(id, OntologeniusConversionRequest.LITERAL)

    def _indexes2Ids(self, indexes, source):
        response = self._call(source, [], indexes)
        if response:
          return response.values_str
        else:
          return None

    def _index2Id(self, index, source):
        response = self._call(source, [], [index])
        if response:
          return response.values_str[0]
        else:
          return None

    def _ids2Indexes(self, ids, source):
        response = self._call(source, ids, [])
        if response:
          return response.values_int
        else:
          return None

    def _id2Index(self, id, source):
        response = self._call(source, [id], [])
        if response:
          return response.values_int[0]
        else:
          return None

    def _call(self, source, values_str, values_int):
        request = OntologeniusConversionRequest(source = source, values_str = values_str, values_int = values_int)
        return self._client.call(request, self._verbose)